File: src/org/scalaconsole/fxui/search/SearchArtifactController.java
package org.scalaconsole.fxui.search;

import com.google.gson.JsonElement;
import javafx.beans.binding.Bindings;
import javafx.beans.property.BooleanProperty;
import javafx.beans.property.SimpleBooleanProperty;
import javafx.event.ActionEvent;
import javafx.fxml.FXML;
import javafx.scene.control.Label;
import javafx.scene.control.ListCell;
import javafx.scene.control.ListView;
import javafx.scene.control.TextField;
import javafx.scene.image.ImageView;
import javafx.scene.input.KeyEvent;
import javafx.scene.layout.FlowPane;
import org.scalaconsole.fxui.SemVersion;

import java.util.Map;

import static javafx.scene.input.KeyCode.ENTER;

public class SearchArtifactController {
    SearchArtifactDelegate delegate;

    @FXML
    ImageView loadingImg;

    @FXML
    Label errorMsg;

    @FXML
    ListView<SemVersion> versionList;

    @FXML
    FlowPane crossBuildsPane;

    @FXML
    ListView<Map.Entry<String, JsonElement>> matchedList;

    @FXML
    ListView<String> selectedVersionList;

    @FXML
    TextField searchBox;

    @FXML
    void onOK(ActionEvent event) {
        SearchArtifactStage window = (SearchArtifactStage) errorMsg.getScene().getWindow();
        window.onOK();
        window.close();
    }

    @FXML
    void onCancel(ActionEvent event) {
        SearchArtifactStage window = (SearchArtifactStage) errorMsg.getScene().getWindow();
        window.close();
    }

    @FXML
    void onEnterInSearchBox(KeyEvent e) {
        if (e.getCode() == ENTER) {
            delegate.onSearch();
        }
    }

    BooleanProperty loading = new SimpleBooleanProperty(false);

    class ArtifactCell extends ListCell<Map.Entry<String, JsonElement>> {
        public ArtifactCell() {
            textProperty().bind(Bindings.createStringBinding(() -> {
                        Map.Entry<String, JsonElement> item = getItem();
                        if (item != null) return item.getKey();
                        else return null;
                    }, itemProperty()
            ));
        }
    }

    class VersionCell extends ListCell<SemVersion> {
        public VersionCell() {
            textProperty().bind(Bindings.createStringBinding(() -> {
                        SemVersion item = getItem();
                        if (item != null) return item.stringPresentation();
                        else return null;
                    }, itemProperty()
            ));
        }
    }

    @FXML
    void initialize() {
        delegate = new SearchArtifactDelegate(this);
        loadingImg.visibleProperty().bind(loading);
        matchedList.setCellFactory(entryListView -> new ArtifactCell());
        matchedList.getSelectionModel().selectedItemProperty().addListener((observableValue, oldEntry, newEntry) -> {
            if (newEntry != null)
                delegate.onSelectArtifact(newEntry);
        });
        versionList.setCellFactory(versionListView -> {
            VersionCell cell = new VersionCell();
            cell.setOnMouseClicked(evt -> {
                if (evt.getClickCount() == 2 && cell.getItem() != null) {
                    delegate.addVersion2Selection(cell.getItem());
                }
            });
            return cell;
        });
        delegate.init();
    }

}

File: src/org/scalaconsole/fxui/search/SearchArtifactController.scala
package org.scalaconsole.fxui.search

import javafx.animation.FadeTransition
import javafx.beans.value.ObservableValue
import javafx.event.ActionEvent
import javafx.fxml.FXML
import javafx.scene.control._
import javafx.scene.input.KeyCode._
import javafx.scene.input.KeyEvent
import javafx.util.Duration
import javax.json.{JsonArray, JsonValue, JsonObject}

import org.scalaconsole.fxui.FxUtil._
import org.scalaconsole.fxui.{SelectedVersionCell, SemVersion, Variables}
import org.scalaconsole.net.MavenIndexerClient

import scala.collection.JavaConverters._

trait SearchArtifactController { self: SearchArtifactStage =>

  @FXML def onOK(event: ActionEvent) {
    import scala.collection.JavaConverters._
    mainStage.addArtifacts(selectedVersionList.getItems.asScala)
    close()
  }

  @FXML def onCancel(event: ActionEvent) {
    close()
  }

  @FXML def onEnterInSearchBox(e: KeyEvent): Unit = {
    if (e.getCode eq ENTER) {
      val keyword = searchBox.getText
      if (keyword.length < 3) {
        animateErrorMsg()
      } else {
        errorMsg.setOpacity(0)
        loading.setValue(true)
        startTask {
          val (exactMatch, otherMatch) = MavenIndexerClient.search(keyword)
          val allMatches = exactMatch ++ otherMatch
          val versions = extractVersions(exactMatch) | extractVersions(otherMatch)
          val versionGroup = new ToggleGroup

          versionGroup.selectedToggleProperty.addListener { (p1: ObservableValue[_ <: Toggle], old: Toggle, selected: Toggle) =>
            val selectedBtn = selected.asInstanceOf[ToggleButton]
            artifactVersions.clear()
            if (selectedBtn.getText == "All") {
              matchedArtifacts.setAll(allMatches.asJavaCollection)
            } else {
              matchedArtifacts.setAll(allMatches.filter { case (k, v) => k.endsWith(selectedBtn.getText)}.asJavaCollection)
            }
          }
          val allBtn = new ToggleButton("All")
          allBtn.setToggleGroup(versionGroup)
          case class VersionButton(btn: ToggleButton, version: SemVersion)
          val versionButtons = for (version <- versions.toSeq; semVersion <- SemVersion(version)) yield {
            val btn = new ToggleButton(version)
            btn.setToggleGroup(versionGroup)
            VersionButton(btn, semVersion)
          }

          onEventThread {
            val children = crossBuildsPane.getChildren
            children.setAll(allBtn)
            children.addAll(versionButtons.toSeq.sortBy(_.version).map(_.btn).asJavaCollection)
            val currentSemVersion = SemVersion(Variables.currentScalaVersion)
            versionButtons
              .find(_.version.fuzzyMatch(currentSemVersion))
              .orElse(versionButtons.find(vb => vb.btn.getText == Variables.currentScalaVersion))
              .map(_.btn)
              .getOrElse(allBtn)
              .setSelected(true)
            loading.setValue(false)
          }
        }
      }
    }
  }

  @FXML def initialize(): Unit = {
    loadingImg.visibleProperty.bind(loading)
    matchedList.setCellFactory((_: ListView[(String, JsonValue)]) => new ArtifactCell())
    matchedList.getSelectionModel.selectedItemProperty.addListener { (p1: ObservableValue[_ <: (String, JsonValue)], oldEntry: (String, JsonValue), newEntry: (String, JsonValue)) =>
      if (newEntry != null)
        onSelectArtifact(newEntry.asInstanceOf[(String, JsonArray)])
    }
    versionList.setCellFactory((_: ListView[SemVersion]) => new VersionCell())
    selectedVersionList.setCellFactory((_: ListView[String]) => new SelectedVersionCell())

    matchedList.setItems(matchedArtifacts)
    versionList.setItems(artifactVersions)
    selectedVersionList.setItems(selectedVersions)
  }

  @FXML def onKeyUp(e: KeyEvent) {
    e.getCode match {
      case ESCAPE =>
        onCancel(null)
      case _ =>
    }
  }

  private def animateErrorMsg() = {
    val ft = new FadeTransition(Duration.millis(200), errorMsg)
    ft.setFromValue(0)
    ft.setToValue(1)
    ft.setCycleCount(5)
    ft.setAutoReverse(true)
    ft.play()
  }

  private def extractVersions(json: Map[String, JsonValue]): Set[String] = {
    json.collect {
      case (SemVersion.R(v), _) => v
    }.toSet
  }

  private def onSelectArtifact(entry: (String, JsonArray)) = {
    val versions = for {
      artifact <- entry._2.asScala
      version <- SemVersion.apply(artifact.asInstanceOf[JsonObject].getString("version"))
    } yield version

    artifactVersions.setAll(versions.toSeq.sorted.asJavaCollection)
  }

  protected def addVersion2Selection(ver: SemVersion) = {
    val dependencyString = s"${matchedList.getSelectionModel.getSelectedItem._1.trim}:${ver.stringPresentation}"
    selectedVersions.add(dependencyString)
  }

}
